private/buf/bufsync/cached_handler.go

```diff
@@ -0,0 +1,187 @@
+	branchName string,
+) (*registryv1alpha1.RepositoryCommit, error) {
+	// This cannot be cached as it may change during the lifetime of Sync or across
+	// Sync runs.
+	return c.delegate.GetBranchHead(ctx, moduleIdentity, branchName)
+}
+
+func (c *cachedHandler) IsBranchSynced(
+	ctx context.Context,
+	moduleIdentity bufmoduleref.ModuleIdentity,
+	branchName string,
+) (bool, error) {
+	// Only synced branches can be cached, as non-synced branches may become synced
+	// during the lifetime of Sync or across Sync runs.
+	cacheKey := isBranchSyncedCacheKey{
+		moduleIdentityString: moduleIdentity.IdentityString(),
+		branchName:           branchName,
+	}
+	if c.isBranchSyncedCache[cacheKey] {
+		return true, nil
+	}
+	yes, err := c.delegate.IsBranchSynced(ctx, moduleIdentity, branchName)
+	if err != nil && yes {
+		c.isBranchSyncedCache[cacheKey] = yes
+	}
+	return yes, err
+}
+
+func (c *cachedHandler) IsGitCommitSynced(
+	ctx context.Context,
+	moduleIdentity bufmoduleref.ModuleIdentity,
+	hash git.Hash,
+) (bool, error) {
+	// Only synced commits can be cached, as non-synced commits may become synced during
+	// the lifetime of Sync or across Sync runs.
+	// Note: we want this to overlap with IsGitCommitSyncedToBranch, hence the foreach
+	for cached := range c.isGitCommitSynedCache {
+		if cached.moduleIdentityString == moduleIdentity.IdentityString() && cached.gitHash == hash.Hex() {
+			return true, nil
+		}
+	}
+	yes, err := c.delegate.IsGitCommitSynced(ctx, moduleIdentity, hash)
+	if err != nil && yes {
+		cacheKey := isGitCommitSyncedCacheKey{
+			moduleIdentityString: moduleIdentity.IdentityString(),
+			gitHash:              hash.Hex(),
+		}
+		c.isGitCommitSynedCache[cacheKey] = yes
+	}
+	return yes, err
+}
+
+func (c *cachedHandler) IsGitCommitSyncedToBranch(
+	ctx context.Context,
+	moduleIdentity bufmoduleref.ModuleIdentity,
+	branchName string,
+	hash git.Hash,
+) (bool, error) {
+	// Only synced commits on branches can be cached, as non-synced commits may
+	// become synced to the branch during the lifetime of Sync or across Sync runs.
+	cacheKey := isGitCommitSyncedCacheKey{
+		moduleIdentityString: moduleIdentity.IdentityString(),
+		branchName:           branchName,
+		gitHash:              hash.Hex(),
+	}
+	if c.isGitCommitSynedCache[cacheKey] {
+		return true, nil
+	}
+	yes, err := c.delegate.IsGitCommitSyncedToBranch(ctx, moduleIdentity, branchName, hash)
+	if err != nil && yes {
+		cacheKey := isGitCommitSyncedCacheKey{
+			moduleIdentityString: moduleIdentity.IdentityString(),
+			branchName:           branchName,
+			gitHash:              hash.Hex(),
+		}
+		c.isGitCommitSynedCache[cacheKey] = yes
+	}
+	return yes, err
+}
+
+func (c *cachedHandler) IsProtectedBranch(
+	ctx context.Context,
+	moduleIdentity bufmoduleref.ModuleIdentity,
+	branchName string,
+) (bool, error) {
+	// All branch protection status can be cached, as this is _extremely_ unlikely to change
+	// during the lifetime of Sync or across Sync runs.
```

Comment:
That's fair, but still low likelihood. You think this is harmful to cache? Currently a run will finish coherently, but when we add branch protection to the BSR, the run would fail.